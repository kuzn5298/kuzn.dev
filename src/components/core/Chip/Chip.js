/* eslint-disable react/require-default-props */
import React from 'react';
import { rgba } from 'polished';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const sizeBlock = css`
    ${({ $size }) => {
        switch ($size) {
            case 'small':
                return css`
                    font-size: 0.75rem;
                    padding: 0.125rem 0.25rem;
                `;
            case 'medium':
                return css`
                    font-size: 1rem;
                    padding: 0.25rem 0.5rem;
                `;
            case 'large':
                return css`
                    font-size: 1.5rem;
                    padding: 0.5rem 1rem;
                `;
            default:
                return undefined;
        }
    }}
`;

const borderColorBlock = css`
    ${({ $borderColor, theme }) => {
        if ($borderColor) {
            switch ($borderColor) {
                case 'primary':
                    return css`
                        border-left: 3px solid ${theme.palette.primary.main};
                    `;
                default:
                    return css`
                        border-left: 3px solid ${$borderColor};
                    `;
            }
        }
        return undefined;
    }}
`;

const colorBlock = css`
    ${({ $color, theme }) => {
        switch ($color) {
            case 'primary':
                return css`
                    background-color: ${rgba(theme.palette.primary.main, 0.75)};
                `;
            case 'secondary':
                return css`
                    background-color: ${theme.palette.common.darkGray};
                `;
            default:
                return css`
                    background-color: ${$color};
                `;
        }
    }}
`;

const ChipStyled = styled.div`
    ${sizeBlock}
    ${colorBlock}
    ${borderColorBlock}

    text-transform: uppercase;
    white-space: nowrap;
`;

const Chip = ({ children = null, color = 'primary', size = 'medium', borderColor, ...props }) => (
    <ChipStyled $color={color} $size={size} $borderColor={borderColor} {...props}>
        {children}
    </ChipStyled>
);

Chip.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    borderColor: PropTypes.oneOfType([PropTypes.oneOf(['primary']), PropTypes.string]),
    color: PropTypes.oneOfType([PropTypes.oneOf(['primary', 'secondary']), PropTypes.string]),
    size: PropTypes.oneOfType([PropTypes.oneOf(['inherit', 'large', 'medium', 'small'])]),
};

export default Chip;
