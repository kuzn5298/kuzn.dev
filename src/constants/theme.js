import { MoonIcon, SunIcon } from '@components/icons';

export const THEME = {
    LIGHT: 'light',
    DARK: 'dark',
};

export const NEXT_THEME = {
    [THEME.DARK]: {
        id: THEME.LIGHT,
        name: 'Light',
        icon: SunIcon,
    },
    [THEME.LIGHT]: {
        id: THEME.DARK,
        name: 'Dark',
        icon: MoonIcon,
    },
};
