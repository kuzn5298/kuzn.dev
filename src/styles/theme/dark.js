import _ from 'lodash';
import defaultTheme from './default';

const darkTheme = {
    palette: {
        background: {
            default: '#1D1C1C',
        },
        text: {
            primary: '#F2F2F2',
        },
        primary: {
            main: '#B3E5FC',
        },
    },
};

export default _.merge({}, defaultTheme, darkTheme);
