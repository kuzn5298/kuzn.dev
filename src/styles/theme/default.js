const defaultTheme = {
    typography: {
        fontFamily: 'Lexend, Arial, Helvetica, sans-serif;',
        fontSize: 16,
    },
};

export default defaultTheme;
