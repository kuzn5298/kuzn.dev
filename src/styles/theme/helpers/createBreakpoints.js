const STEP = 5 / 100;

const createBreakpoints = (values, unit = 'px') => {
    const keys = Object.keys(values);

    const up = (key) => {
        const value = values[key] ?? key;
        return `@media (min-width: ${value}${unit})`;
    };

    const down = (key) => {
        const value = values[key] ?? key;
        return `@media (max-width: ${value - STEP}${unit})`;
    };

    return {
        keys,
        values,
        up,
        down,
        unit,
    };
};

export default createBreakpoints;
