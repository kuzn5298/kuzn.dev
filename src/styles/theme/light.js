import _ from 'lodash';
import defaultTheme from './default';

const lightTheme = {
    palette: {
        background: {
            default: '#fff',
        },
        text: {
            primary: '#000',
        },
        primary: {
            main: '#B3E5FC',
        },
    },
};

export default _.merge({}, defaultTheme, lightTheme);
