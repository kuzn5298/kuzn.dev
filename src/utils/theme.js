import { THEME, DEFAULT_THEME_ID, THEME_LOCAL_STORAGE, IS_BROWSER } from '@constants';
import storage from '@libs/storage';

export const isValidThemeId = (themeId) => Object.values(THEME).includes(themeId);

export const setThemeIdToLocalStorages = (themeId) => {
    storage.set(THEME_LOCAL_STORAGE, themeId);
};

export const getThemeIdFromLocalStorages = () => {
    const themeId = storage.get(THEME_LOCAL_STORAGE);
    return isValidThemeId(themeId) ? themeId : null;
};

export const getSystemTheme = () => {
    const isDarkTheme = IS_BROWSER
        ? window.matchMedia('(prefers-color-scheme: dark)')?.matches
        : false;
    return isDarkTheme ? THEME.DARK : null;
};

export const getUsedThemeId = () => {
    const localTheme = getThemeIdFromLocalStorages();
    const systemTheme = getSystemTheme();
    return localTheme ?? systemTheme ?? DEFAULT_THEME_ID;
};
