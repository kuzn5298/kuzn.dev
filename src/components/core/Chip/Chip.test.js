import React from 'react';
import { rgba } from 'polished';
import { lightTheme } from '@styles/theme';
import { wrapWithTheme } from '@tests/helpers';
import { render, screen } from '@testing-library/react';
import Chip from './Chip';

describe('<Chip />', () => {
    it('should render component with default props', () => {
        render(wrapWithTheme(<Chip>Default Chip</Chip>));
        const chipElement = screen.getByText(/default chip/i);

        expect(chipElement).toBeInTheDocument();
        expect(chipElement).toHaveStyle({
            backgroundColor: rgba(lightTheme.palette.primary.main, 0.75),
            fontSize: '1rem',
            padding: '0.25rem 0.5rem',
        });
    });

    it('should render a correct chip color', () => {
        const { getByTestId } = render(
            wrapWithTheme(
                <>
                    <Chip color="primary" data-testid="primary-chip">
                        Test Chip
                    </Chip>
                    <Chip color="secondary" data-testid="secondary-chip">
                        Test Chip
                    </Chip>
                </>,
                lightTheme
            )
        );

        expect(getByTestId('primary-chip')).toHaveStyle(
            `backgroundColor: ${rgba(lightTheme.palette.primary.main, 0.75)}`
        );
        expect(getByTestId('secondary-chip')).toHaveStyle(
            `backgroundColor: ${lightTheme.palette.common.darkGray}`
        );
    });

    it('should render a correct chip size', () => {
        const { getByTestId } = render(
            wrapWithTheme(
                <>
                    <Chip size="small" data-testid="small-chip">
                        Test Chip
                    </Chip>
                    <Chip size="medium" data-testid="medium-chip">
                        Test Chip
                    </Chip>
                    <Chip size="large" data-testid="large-chip">
                        Test Chip
                    </Chip>
                    <Chip size="inherit" data-testid="inherit-chip">
                        Test Chip
                    </Chip>
                </>,
                lightTheme
            )
        );

        expect(getByTestId('small-chip')).toHaveStyle({
            fontSize: '0.75rem',
            padding: '0.125rem 0.25rem',
        });
        expect(getByTestId('medium-chip')).toHaveStyle({
            fontSize: '1rem',
            padding: '0.25rem 0.5rem',
        });
        expect(getByTestId('large-chip')).toHaveStyle({
            fontSize: '1.5rem',
            padding: '0.5rem 1rem',
        });
        expect(getByTestId('inherit-chip')).toHaveStyle({
            fontSize: 'inherit',
            padding: 'inherit',
        });
    });

    it('should render a correct chip border color', () => {
        const { getByTestId } = render(
            wrapWithTheme(
                <Chip borderColor="primary" data-testid="primary-chip">
                    Test Chip
                </Chip>,
                lightTheme
            )
        );

        expect(getByTestId('primary-chip')).toHaveStyle(
            `borderLeftColor: ${lightTheme.palette.primary.main};`
        );
    });

    it('should render a correct chip custom props', () => {
        const { getByTestId } = render(
            wrapWithTheme(
                <Chip color="#ff0000" borderColor="#00ff00" data-testid="custom-chip">
                    Test Chip
                </Chip>
            )
        );

        expect(getByTestId('custom-chip')).toHaveStyle({
            backgroundColor: '#ff0000',
            borderLeftColor: '#00ff00',
        });
    });

    it('should render a chip with different props (snapshot test)', () => {
        const { asFragment } = render(
            wrapWithTheme(
                <Chip color="secondary" size="small" borderColor="primary">
                    Test Chip
                </Chip>
            )
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
