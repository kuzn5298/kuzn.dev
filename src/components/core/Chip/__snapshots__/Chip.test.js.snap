// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip Component correct custom props (snapshot test) 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 0.75rem;
  padding: 0.125rem 0.25rem;
  background-color: rgba(158, 158, 158, 0.2);
  border-left: 3px solid #03a9f4;
  text-transform: uppercase;
  white-space: nowrap;
}

<div
    class="c0"
  >
    Custom Chip
  </div>
</DocumentFragment>
`;
